components/tree-grid/index.jsx

```diff
@@ -0,0 +1,293 @@
+			flattenedNodes,
+			focusedNodeIndex,
+		};
+
+		this.generatedId = shortid.generate();
+	}
+
+	componentWillReceiveProps(nextProps) {
+		this.setState({
+			flattenedNodes: this.flattenTree({
+				nodes: this.props.getNodes({ nodes: nextProps.nodes }),
+				expanded: true,
+			}).slice(1),
+		});
+	}
+
+	/**
+	 * Get the TreeGrid's HTML id. Generate a new one if no ID present.
+	 */
+	getId() {
+		return this.props.id || this.generatedId;
+	}
+
+	flattenTree = (root, treeIndex = '', firstLevel = true) => {
+		if (!root.nodes) {
+			return [{ node: root, treeIndex }];
+		}
+		let nodes = [{ node: root, treeIndex }];
+		if (root.expanded) {
+			for (let index = 0; index < root.nodes.length; index += 1) {
+				const curNode = firstLevel
+					? root.nodes[index]
+					: this.props.getNodes(root)[index];
+				nodes = nodes.concat(
+					this.flattenTree(
+						curNode,
+						treeIndex ? `${treeIndex}-${index}` : `${index}`,
+						false
+					)
+				);
+			}
+		}
+		return nodes;
+	};
+
+	handleFocus = (event, data) => {
+		this.setState({
+			focusedNodeIndex: data.treeIndex,
+		});
+	};
+
+	handleSelection = (event, data) => {
+		this.props.onSelect(event, data);
+	};
+
+	handleExpansion = (event, data) => {
+		this.props.onExpand(event, data);
+	};
+
+	render() {
+		const assistiveText = assign(
+			{},
+			defaultProps.assistiveText,
+			this.props.assistiveText
+		);
+
+		const columns = [];
+		React.Children.forEach(this.props.children, (child) => {
+			if (child && child.type.displayName === TreeGridColumn.displayName) {
+				const { children, ...columnProps } = child.props;
+
+				const props = assign({}, this.props);
+				delete props.children;
+				assign(props, columnProps);
+
+				columns.push({
+					props,
+					dataTableProps: this.props,
+				});
+			}
+		});
+
+		return (
+			<table
+				id={this.getId()}
+				aria-readonly="true"
+				aria-multiselectable={this.props.selectRows === 'multiple'}
+				role="treegrid"
+				className={classNames(
+					'slds-table',
+					'slds-table_edit',
+					'slds-table_fixed-layout',
+					'slds-tree slds-table_tree',
+					{ 'slds-table_bordered': !this.props.isBorderless },
+					{ 'slds-table_header-hidden': this.props.isHeadless },
+					this.props.className
+				)}
+			>
+				{this.props.isHeadless ? null : (
+					<thead>
+						<tr className="slds-line-height_reset">
+							{this.props.selectRows === 'multiple' ? (
+								<th
+									className="slds-text-align_right"
+									scope="col"
+									style={{ width: '3.5rem' }}
+								>
+									<span
+										id="column-group-header"
+										className="slds-assistive-text"
+									>
+										{assistiveText.selectRow}
+									</span>
+									<div
+										className="slds-th__action slds-th__action_form"
+										tabIndex="-1"
+									>
+										<Checkbox
+											assistiveText={{
+												label: assistiveText.selectAll,
+											}}
+											name="options"
+											// indeterminate={
```

Comment:
I think I might try an reducer function passed down that manipulates a `isSelectAll` state in `index.js` https://www.w3schools.com/jsref/jsref_reduce.asp

pseudo code: `selectAll = true if all rows rendered are selected` so on each row render you'd run the function and update the parent's checkbox state.

The main issue is then what do we do about non-rendered rows. Although I think we we pass an `onSelectAll` function out, it's up to the dev to update their data, so I'm ok, with select all checkbox only reflecting real data state. Does that makes sense?

In DataTable we did this, but it was an array.
```
		const indeterminateSelected =
			canSelectRows && numRows !== numSelected && numSelected !== 0;
```

@kevinparkerson What do you think?

Reply:
I'm hesitant to `this.flattenTree` but maybe we should re-use that function since it already walks the tree.